python/cuspatial/cuspatial/core/binpreds/binpreds.py

```diff
@@ -168,23 +185,71 @@ def _op(self, lhs, points):
             raise TypeError(
                 "`.contains` can only be called with polygon series."
             )
+        if self._should_use_quadtree():
+            point_result = quadtree_contains_properly(
+                points,
+                lhs,
+            )
+        else:
+            point_result = byte_limited_contains_properly(points, lhs)
+        return point_result
 
-        # call pip on the three subtypes on the right:
-        point_result = contains_properly(
-            points.x,
-            points.y,
-            lhs.polygons.part_offset,
-            lhs.polygons.ring_offset,
-            lhs.polygons.x,
-            lhs.polygons.y,
+    def _postprocess_quadtree_result(self, point_indices, point_result):
+        # If there are more than 31 polygons, the point indices are
+        # returned as a dataframe with two columns: part_index and
+        # point_index.
+
+        # This complex block of code is to create a dataframe that
+        # contains the polygon index and the point index for each
+        # point in the polygon. Quadtree pip returns the _part_index_
+        # of the polygon, not the polygon index.
+        rings_to_parts = cp.array(self.lhs.polygons.part_offset)
+        part_sizes = rings_to_parts[1:] - rings_to_parts[:-1]
+        parts_map = cudf.Series(
+            cp.arange(len(part_sizes)), name="part_index"
+        ).repeat(part_sizes)
+        parts_df = parts_map.reset_index(drop=True).reset_index()
+        # Mapping of parts to polygons
+        parts_to_geoms = cp.array(self.lhs.polygons.geometry_offset)
+        geometry_sizes = parts_to_geoms[1:] - parts_to_geoms[:-1]
+        geometry_map = cudf.Series(
+            cp.arange(len(geometry_sizes)), name="polygon_index"
+        ).repeat(geometry_sizes)
+        geom_df = geometry_map.reset_index(drop=True)
+        geom_df.index.name = "part_index"
```

Comment:
Put this complex block of code in a clearly named function with clearly named parameters.

Reply:
Great idea, this is easy to do and exposes what the function does in a way that helps majorly with comprehension. +1 for adding complexity this time for sure.

Reply:
done